src/lib/dialog/dialog-container.ts

```diff
@@ -23,23 +27,52 @@ import {MdDialogContentAlreadyAttachedError} from './dialog-errors';
   },
   encapsulation: ViewEncapsulation.None,
 })
-export class MdDialogContainer extends BasePortalHost {
+export class MdDialogContainer extends BasePortalHost implements OnDestroy {
   /** The portal host inside of this container into which the dialog content will be loaded. */
   @ViewChild(PortalHostDirective) _portalHost: PortalHostDirective;
 
+  /** The directive that traps and manages focus within the dialog. */
+  @ViewChild(FocusTrap) _focusTrap: FocusTrap;
+
+  /** Element that was focused before the dialog was opened. Save this to restore upon close. */
+  private _elementFocusedBeforeDialogWasOpened: Element = null;
+
   /** The dialog configuration. */
   dialogConfig: MdDialogConfig;
 
+  constructor(private _ngZone: NgZone) {
+    super();
+  }
+
   /** Attach a portal as content to this dialog container. */
   attachComponentPortal<T>(portal: ComponentPortal<T>): ComponentRef<T> {
     if (this._portalHost.hasAttached()) {
       throw new MdDialogContentAlreadyAttachedError();
     }
 
-    return this._portalHost.attachComponentPortal(portal);
+    let attachResult = this._portalHost.attachComponentPortal(portal);
+
+    // If were to attempt to focus immediately, then the content of the dialog would not yet be
+    // ready in stances where change detection has to run first. To deal with this, we simply
+    // wait for the microtask queue to be empty.
+    this._ngZone.onMicrotaskEmpty.first().subscribe(() => {
+      this._elementFocusedBeforeDialogWasOpened = document.activeElement;
+      this._focusTrap.focusFirstTabbableElement();
+    });
+
+    return attachResult;
   }
 
   attachTemplatePortal(portal: TemplatePortal): Map<string, any> {
     throw Error('Not yet implemented');
   }
+
+  ngOnDestroy() {
+    // When the dialog is destroyed, return focus to the element that originally had it before
+    // the dialog was opened. Wait for the DOM to finish settling before changing the focus so
+    // that it doesn't end up back on the <body>.
+    this._ngZone.onMicrotaskEmpty.first().subscribe(() => {
+      (this._elementFocusedBeforeDialogWasOpened as HTMLElement).focus();
```

Comment:
I didn't know the `as` syntax existed until recently; I do this it is a bit cleaner and more in keeping with TypeScript's normally putting the type at the end. We can chat about it when we're all back in the office.